### PR TITLE
Require a restart after changing hardware rendering

### PR DESCRIPTION
**Pull request by HaasJona**
Require a restart after changing hardware rendering setting, decouple setting from active status

### CODE REVIEW

File: src/windows/options.c [hunk 1]
```diff
@@ -490,7 +490,7 @@ static void window_options_mouseup(rct_window *w, int widgetIndex)
 			break;
 		case WIDX_HARDWARE_DISPLAY_CHECKBOX:
 			gConfigGeneral.hardware_display ^= 1;
-			platform_refresh_video();
+			window_error_open(5554, STR_NONE);
```
Comment (IntelOrca) [inline src/windows/options.c]: Please use string constants, add 5554 to string_ids.h.

**Comment by IntelOrca** [comment]
A restart is not necessary on Windows, please use #ifndef WIN32

**Comment by HaasJona** [comment]
> A restart is not necessary on Windows, please use #ifndef WIN32

Hmm, can we guarantee that it always work in Windows? According to the API documentation I saw this seems to an unsupported operation across systems. Another approach might be to try to switch on the fly and if that fails then don't crash :wink: and instead prompt the user to restart.

> Please use string constants, add 5554 to string_ids.h.

Will do, thanks for the hint.

**Comment by IntelOrca** [comment]
If it fails maybe, but we have had no reports yet of it failing except on linux.

**Comment by vzybilly** [comment]
I remember a good deal of issues when switching... I'm not sure how those
issues were fixed but most of that could probably be removed with having a
restart...

**Comment by pizza2004** [comment]
I agree that if the documentation says you're not allowed to perform the function without restarting then we should be enforcing that across all systems, whether it happens to work on one or not.  It's bad coding practice to use a library in an unsupported way because it happens to work.  If SDL updated and it happened to break this functionality on Windows, since it's not a supported function, then we'd need to recode it anyway.  Plus, I'd be unsurprised if some of the issues people have had with hardware acceleration come from not restarting.  I know I've never thought to restart after changing the setting, and I've always seen issues when using it.  It's tough to tell what sort of problems it could be causing.

**Comment by IntelOrca** [comment]
I'd disagree.
Also resetting the window should be the minimal necessary not restarting the game. Even if we have to dispose the entire SDL2 platform layer (which seems silly).

**Comment by vzybilly** [comment]
restarting the window would be enough of a restart for me... I'm just worried about code around keeping track of the first window, causing issues there.

**Comment by IntelOrca** [comment]
Well there should only be one global reference to the window. So it should just be a matter of calling the window dispose routine and running the initialise window routine. Those routines must only do window stuff, otherwise it will end up re-initialising stuff it doesn't need to.

**Comment by HaasJona** [comment]
I fixed your comments on the patch. It will switch on the fly for Windows and ask for a restart on Linux. If creating a hardware renderer fails, it will switch back to software rendering and output a warning.

> Well there should only be one global reference to the window. So it should just be a matter of calling the window dispose routine and running the initialise window routine. Those routines must only do window stuff, otherwise it will end up re-initialising stuff it doesn't need to.

Well, that didn't work for me, switching from hardware to software rendering caused the screen to not update anymore in about 90% of the cases. You are welcome to try, though.

**Comment by HaasJona** [comment]
What's the status of this? I would need to rebase and update the string I'd before this can be merged.

Our if you want to try fixing this by recreating the window instead I could scrap this.

**Comment by IntelOrca** [comment]
Change it to re-create the window, and only for linux.

**Comment by HaasJona** [comment]
> or at least just make this PR only apply to linux.

Windows behaviour is unchanged in this PR currently. Just a little refactoring.

I can try implementing recreating the window next weekend probably, when I have more time. But I will probably need some help as I don't understand all the stuff completely.

**Comment by IntelOrca** [comment]
OK, rebase this and I will merge.

**Comment by HaasJona** [comment]
Okay, rebased. Please make sure I didn't broke anything in Windows because I'm currently only able to test Linux. But as far as I can see, nothing should have changed.

**Pull request merged**